Profile screen

randrade13 opened the pull request:
Work to be merged - Profile Screen:

-Creation of profile screen w/ tab bar configured
-Profile screen displays user's profile picture and username
-Users can change profile picture using camera (or camera roll if camera not available) and changes saved to Parse server
-Creation of completed tasks table view within profile screen with infinite scroll functionality.
-Pull to refresh functionality for completed tasks table view.
-Users can "un-complete" task by tapping checkbox and refreshing.
-Auto-layout for Profile screen and table completed
-Changes Infinite Scroll Framework: users can setup table with single command, infinite scroll framework now handles activity indicator / refresh control as long as fetch data function implemented by user and proper outlets are made.

Commit:
Created profile screen, linked to tab bar, basic functionality added: display profile picture and username

Commit:
User can now change profile picture (camera if phone, or camera roll if unavailable). Changes are upadted on Parse server.

Commit:
Created completed task table view within Profile Screen complete with infinite scroll functionality as well as ability to 'uncomplete' task which will return to incomplete task list. Autolayout not currently set, will do shortly.

Commit:
Added autolayout to profile view and its table. Profile view essential components now finished.

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+    [self refreshCompleteTaskList];
+}
+
+- (void)setupProfile{
```
nit: add space between last letter and `{`
```suggestion
- (void)setupProfile {
```

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+    [self loadUserProfilePicture];
+}
+
+-(void)loadUserProfilePicture{
```
same here
```suggestion
-(void)loadUserProfilePicture {
```

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+    return newImage;
+}
+
+- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
```
is this the only place where we are using pickercontroller?

lmag94 replied:
I mean, is this the only view controller?

randrade13 replied:
Currently yes, but if we end up implementing photo-based task verification it might be used again.

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+    [self presentViewController:imagePickerVC animated:YES completion:nil];
+}
+
+- (void)refreshCompleteTaskList{
```
nit: spacing
```suggestion
- (void)refreshCompleteTaskList {
```

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+    self.completedTasksListInfiniteScrollView.infiniteScrollDelegate = self;
+    [self setupProfile];
+    [self fetchData];
+    [self refreshCompleteTaskList];
```
we should start animating the activity indicators before fetching the data, let's say fetchData is fast and it finishes right before we call startAnimating, the activity indicator will keep animating until next reload.

randrade13 replied:
This has been fixed

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+    }];
+}
+
+- (void)loadMoreData{
```
nit: spacing
```suggestion
- (void)loadMoreData {
```

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+    self.latestDate = latestTask.createdAt;
+    
+    [BorbParseManager loadMoreCompleteTasksOfUser:User.currentUser.username withLaterDate:self.latestDate WithCompletion:^(NSMutableArray *posts) {
+        if ([posts count] > 0){
```
nit: spacing
```suggestion
        if ([posts count] > 0) {
```

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+            [self.completedTasksListInfiniteScrollView.tableView reloadData];
+            self.completedTasksListInfiniteScrollView.isMoreDataLoading = false;
+        } else {
+            self.completedTasksListInfiniteScrollView.isMoreDataLoading = true;
```
why do we set this to true?

randrade13 replied:
So that you can't pull down to load more data when there is none, instead giving the user a smoother bounce effect at bottom.

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.m:
```diff
@@ -106,6 +106,42 @@ + (void)loadMoreIncompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate
 
++ (void)fetchCompleteTasksOfUser:(NSString *)username WithCompletion:(void (^)(NSMutableArray *))completion {
+    PFQuery *query = [PFQuery queryWithClassName:QUERY_TASK_NAME];
+    query.limit = 20;
```
let's make this a constant? We are using the same number multiple times

randrade13 replied:
Yea that's true; will do.

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.m:
```diff
@@ -106,6 +106,42 @@ + (void)loadMoreIncompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate
     
 }
 
++ (void)fetchCompleteTasksOfUser:(NSString *)username WithCompletion:(void (^)(NSMutableArray *))completion {
```
```suggestion
+ (void)fetchCompleteTasksOfUser:(NSString *)username withCompletion:(void (^)(NSMutableArray *))completion {
```

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.h:
```diff
@@ -32,8 +32,14 @@ NS_ASSUME_NONNULL_BEGIN
 
 + (void)loadMoreIncompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate *)date WithCompletion:(void (^)(NSMutableArray *))completion;
 
++ (void)fetchCompleteTasksOfUser:(NSString *)username WithCompletion:(void (^)(NSMutableArray *))completion;
```
Should we add an error parameter to the completion block?
```suggestion
+ (void)fetchCompleteTasksOfUser:(NSString *)username withCompletion:(void (^)(NSMutableArray *))completion;
```

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.m:
```diff
@@ -106,6 +106,42 @@ + (void)loadMoreIncompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate
+        if (posts != nil) {
+            completion([NSMutableArray arrayWithArray:posts]);
+        } else {
+            NSLog(@"%@", error.localizedDescription);
```
we should call completion with error

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.m:
```diff
@@ -106,6 +106,42 @@ + (void)loadMoreIncompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate
+        if (posts != nil) {
+            completion([NSMutableArray arrayWithArray:posts]);
+        } else {
+            NSLog(@"%@", error.localizedDescription);
```
same here

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.m:
```diff
@@ -106,6 +106,42 @@ + (void)loadMoreIncompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate
+    }];
+}
+
++ (void)loadMoreCompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate *)date WithCompletion:(void (^)(NSMutableArray *))completion {
```
```suggestion
+ (void)loadMoreCompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate *)date withCompletion:(void (^)(NSMutableArray *))completion {
```

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.h:
```diff
@@ -32,8 +32,14 @@ NS_ASSUME_NONNULL_BEGIN
 
++ (void)fetchCompleteTasksOfUser:(NSString *)username WithCompletion:(void (^)(NSMutableArray *))completion;
+
++ (void)loadMoreCompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate *)date WithCompletion:(void (^)(NSMutableArray *))completion;
```
```suggestion
+ (void)loadMoreCompleteTasksOfUser:(NSString *)username withLaterDate:(NSDate *)date withCompletion:(void (^)(NSMutableArray *))completion;
```

lmag94 commented on borbskeeper/ParseManager/BorbParseManager.m:
```diff
@@ -128,4 +164,15 @@ + (void)signOutUser{
+    if (!imageData) {
+        return nil;
+    }
+    return [PFFileObject fileObjectWithName:@"image.png" data:imageData];
```
why are we using "image.png", is it safe if we get multiple images?

randrade13 replied:
I don't think it currently matters if we use "image.png" because the only way we are accessing images is when they are associated with something (i.e. user-> profile image). I would also imagine in the future if we were to implement photo verification for tasks posted to the social the images would be associated to every task/post. At least in the sense it's 'safe' because each associated object has a property for a single UIImage, not multiple so its overwritten when adding a new one. There is also no situation where we need to extract all images or filtering queries by name. But I do see why that could be a problem in other applications.

lmag94 commented on borbskeeper/Features/Profile/ProfileViewController.m:
```diff
@@ -0,0 +1,162 @@
+- (void)refreshCompleteTaskList{
+    self.refreshControl = [[UIRefreshControl alloc] init];
+    [self.refreshControl addTarget:self action:@selector(fetchData) forControlEvents:UIControlEventValueChanged];
+    [self.completedTasksListInfiniteScrollView.tableView insertSubview:self.refreshControl atIndex:0];
```
We should probably let the `InifiniteScrollView` do this for us.

randrade13 replied:
Yea, I think so too.

lmag94 reviewed (changes requested):
Good job, lots of progress and I see we are using constants and keeping the code clean and organized. I just have a couple of comments regarding style and some things that we should change to make our infinite scroll view more generic and avoid repeating code.

randrade13 committed:
Update borbskeeper/Features/Profile/ProfileViewController.m

stylistic changes.

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/Features/Profile/ProfileViewController.m

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/Features/Profile/ProfileViewController.m

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/Features/Profile/ProfileViewController.m

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/Features/Profile/ProfileViewController.m

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/ParseManager/BorbParseManager.m

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/ParseManager/BorbParseManager.h

Maybe, we haven't decided yet how we want to handle alerts for network errors / inability to fetch data

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/ParseManager/BorbParseManager.m

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

randrade13 committed:
Update borbskeeper/ParseManager/BorbParseManager.h

Co-Authored-By: Luis Mariano <lmag94@gmail.com>

Commit:
Minor fixes to profile screen, update to infinte scroll framework, refresh control & activity indicator handled within the Infinite scroll view as long as fetch data function is defined in the view controller. table setup now also done within infinite scroll view with setup table function.
